### PR TITLE
Separate DeploymentPlan creation from its execution

### LINKED ISSUE #318: Separate DeploymentPlan creation from its execution
### Issue

The `deploy` command automatically execute a deployment plan. Each operation result is inserted in the database sequentially. When an operation fails, the following operations are skipped and are not persisted in the database. So, we can't be sure that resuming a failed deployment effectively execute the initially planed operations.

### Proposition

We should be able to separate the planification from the execution. Hence, to persist the whole deployment plan in the database before launching it.

- The `plan` command must save a deployment plan with all of its operations in the database.
- The `deploy` command should only launch the planed deployment.
- If no deployment is planed, the command raises an error.
- We can't plan 2 deployments in parallel. If more than 1 planed deployment exists in the database, an error is raised when doing the deploy.
- Creating a new plan or modifying one will override the initial one.
- The `resume` deployment mode starts from the failed operation if the last deployment is a failure.
- The `reconfigure` deployment mode can be treated later (see #308)
- Planed deployment should be edited (to do later)

### Implementation

We will need to:

- [x] Add new value to the status enum
- [x] Update the database schema
- [x] Write the whole deployment plan in the database before execution
- [x] Rewrite the execution Iterator
- [x] Rewrite the logic of the plan, resume and deploy commands

#### Database schema

Add an `order` column to the `operation_logs`. Maybe rename the `operation_log` to `operation` table.

#### Status enum

The following Operation status can be considered:

- **PLANNED**, deployment is not launched, operation can be modified.
- **PENDING**, deployment is launched, operation is waiting to be executed.
- **RUNNING**, operation is being executed.
- **SUCCESS**/**FAILURE**, operation succeeded/failed.
- **HELD**, the operation hasn't been executed (a previous one has been paused or failed).

Lifecycle example:

- Planification

  ```
  PLANNED
  PLANNED
  PLANNED
  PLANNED
  ```

- Execution

  ```
  SUCCESS
  RUNNING
  PENDING
  PENDING
  ```

- In case of a failure

  ```
  SUCCESS
  FAILURE
  HELD
  HELD
  ```

### LINKED ISSUE #323: Can't run the same operation multiple time in the same deployment
When doing a deployment in **run** mode (a deployment from an arbitrary list of operations, different from dag mode), we should be able to run the same operation multiple times. At least, it is allowed by the lib/server/ui.

However, the primary key of the `operation_log` table is composed of the `deployment_id` and `operation` name. Hence, we're not able to execute the same operation more than once in the same deployment.

This would probably be fixed by #318

### PR DESCRIPTION
**Pull request by PaulFarault**
#### Which issue(s) this PR fixes

Fixes #318 

Also fixes #323 as operation name is not the primary key anymore.

#### Additional comments

This PR will be incremental.

#### Agreements

- [X] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
- [X] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

**Comment by PaulFarault** [comment]
This PR is now ready for review.

Some test are broken as `DeploymentPlan` and `DeploymentRunner` have been edited. As we need this feature quickly I'll open an issue to fix them once it is merged.

File: tdp/cli/commands/plan/__init__.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])
+
+
+@click.group(
+    context_settings=CONTEXT_SETTINGS,
+    invoke_without_command=True,
+    short_help="Generate a deployment plan.",
+)
+@env
```
Comment (rpignolet) [inline tdp/cli/commands/plan/__init__.py:19]: Why do you need to add `@env`. It is already done by the parent command no ?

File: tdp/cli/commands/deploy.py [hunk 2]
```diff
@@ -86,72 +38,59 @@ def deploy(
...
+        if previous_deployment_log is None:
+            raise click.ClickException(
+                "No previous deployment found, please run `tdp plan` first."
             )
-        except Exception as e:
-            raise click.ClickException(str(e)) from e
+        deployment_plan = DeploymentPlan.from_deployment_log(
+            collections, previous_deployment_log
+        )
+        print(f"COUNT: {len(deployment_plan.deployment_log.operations)}")
```
[the remark this answers is not in the document]
Comment (PaulFarault) [reply tdp/cli/commands/deploy.py]: Yep, I removed it.

File: tdp/cli/commands/plan/__init__.py (hunk 1 above)
Comment (PaulFarault) [reply tdp/cli/commands/plan/__init__.py]: True, I though I had to put it again but that's not useful!

File: tdp/cli/__main__.py [hunk 3]
```diff
@@ -41,8 +32,8 @@ def tdp(env):
 tdp.add_command(init)
 tdp.add_command(nodes)
 tdp.add_command(playbooks)
-tdp.add_command(reconfigure)
-tdp.add_command(resume)
-tdp.add_command(run)
 tdp.add_command(service_versions)
 tdp.add_command(validate)
+tdp.add_command(deploy)
```
[the remark this answers is not in the document]
Comment (PaulFarault) [reply tdp/cli/__main__.py]: Furthermore, this line appeared twice, I removed it.

**Comment by PaulFarault** [comment]
Seems like the first plan log a deployment id that is `None`. Should I retrieve the id from the database or modify the log to stop showing the id?

File: tdp/cli/commands/deploy.py [hunk 4]
```diff
@@ -86,72 +38,58 @@ def deploy(
...
-            click.echo(f"Deploying TDP")
-        try:
-            deployment_plan = DeploymentPlan.from_dag(
-                dag,
-                sources=sources,
-                targets=targets,
-                filter_expression=filter,
-                filter_type=filter_type,
-                restart=restart,
+        previous_deployment_log = get_planned_deployment_log(session)
```
Comment (rpignolet) [inline tdp/cli/commands/deploy.py:59]: I think the variable name can be better.
```suggestion
        planned_deployment_log = get_planned_deployment_log(session)
```

File: tdp/cli/commands/deploy.py [hunk 5]
```diff
@@ -86,72 +38,58 @@ def deploy(
...
+            # Update deployment log to RUNNING
+            session.merge(deployment_iterator.deployment_log)
             session.commit()
-            for operation_log, service_component_log in deployment_iterator:
+            for index, (operation_log, service_component_log) in enumerate(
+                deployment_iterator
+            ):
                 if operation_log is not None:
-                    session.add(operation_log)
+                    operation_log.order = index
```
Comment (rpignolet) [inline tdp/cli/commands/deploy.py:79]: Why the order is set here ? The `OperationLog` returned by the iterator should be ready to put in the database without modification.

The order is already defined when the operation is store in the database by the `plan`.

File: tdp/cli/utils.py [hunk 6]
```diff
@@ -72,13 +72,23 @@ def dry(func: FC) -> FC:
     )(func)
 
 
-def mock_deploy(fun: FC) -> FC:
+def env(func: FC) -> FC:
```
Comment (rpignolet) [inline tdp/cli/utils.py:75]: This option is used once no ? If this is use once, no need to put it here.

File: tdp/core/deployment/deployment_plan.py [hunk 7]
```diff
@@ -101,22 +137,15 @@ def from_reconfigure(dag, cluster_variables, service_component_deployed_version)
             restart=True,
             filter_expression=r".+_(config|(re|)start)",
             filter_type=FilterTypeEnum.REGEX,
+            state=DeploymentStateEnum.PLANNED,
```
Comment (rpignolet) [inline tdp/core/deployment/deployment_plan.py:140]: I do not see `state` argument for `DeploymentPlan.from_dag` and the state is not needed, the `DeploymentPlan.from_dag` function already put a `PLANNED` state.

File: tdp/core/deployment/deployment_plan.py [hunk 8]
```diff
@@ -164,8 +193,25 @@ def from_failed_deployment(dag, deployment_log):
...
         )
-        return DeploymentPlan(remaining_operations, deployment_args)
+        return DeploymentPlan(remaining_operations, new_deployment_log)
+
+    @staticmethod
+    def from_deployment_log(
+        collections: Collections, deployment_log: DeploymentLog
+    ) -> "DeploymentPlan":
+        """Generate a DeploymentPlan from a DeploymentLog."""
```
Comment (rpignolet) [inline tdp/core/deployment/deployment_plan.py:207]: I think the doc should say that only a "planned" state for `DeploymentLog` is accepted.
```suggestion
        """Generate a DeploymentPlan from a "planned" DeploymentLog."""
```

File: tdp/cli/__main__.py [hunk 9]
```diff
@@ -41,8 +39,7 @@ def tdp(env):
 tdp.add_command(init)
 tdp.add_command(nodes)
 tdp.add_command(playbooks)
-tdp.add_command(reconfigure)
-tdp.add_command(resume)
-tdp.add_command(run)
 tdp.add_command(service_versions)
 tdp.add_command(validate)
+
+tdp.add_command(plan)
```
[the remark this answers is not in the document]
Comment (PaulFarault) [reply tdp/cli/__main__.py:45]: I left it appart as it is a `group` and not a standard `command`.

I can put it with the others if you think it's weird.

[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/cli/__main__.py:45]: I think you can put it with the others.

File: tdp/cli/utils.py (hunk 6 above)
Comment (PaulFarault) [reply tdp/cli/utils.py:75]: True, this fonction is not even used anymore, I forgot to removed it.

File: tdp/cli/__main__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (PaulFarault) [reply tdp/cli/__main__.py]: Done

File: tdp/cli/commands/deploy.py (hunk 4 above)
Comment (PaulFarault) [reply tdp/cli/commands/deploy.py]: Done

File: tdp/cli/utils.py (hunk 6 above)
Comment (PaulFarault) [reply tdp/cli/utils.py]: Done

File: tdp/core/deployment/deployment_plan.py (hunk 7 above)
Comment (PaulFarault) [reply tdp/core/deployment/deployment_plan.py]: Done

File: tdp/core/deployment/deployment_plan.py (hunk 8 above)
Comment (PaulFarault) [reply tdp/core/deployment/deployment_plan.py]: Done

**Comment by PaulFarault** [comment]
Rebased on master

**Comment by PaulFarault** [comment]
Fixed license headers

File: tdp/cli/commands/deploy.py [hunk 10]
```diff
@@ -86,72 +38,57 @@ def deploy(
...
-                dag,
-                sources=sources,
-                targets=targets,
-                filter_expression=filter,
-                filter_type=filter_type,
-                restart=restart,
+        planned_deployment_log = get_planned_deployment_log(session)
+        if planned_deployment_log is None:
+            raise click.ClickException(
+                "No previous deployment found, please run `tdp plan` first."
```
Comment (rpignolet) [inline tdp/cli/commands/deploy.py:62]: Also need to be changed here.
```suggestion
                "No planned deployment found, please run `tdp plan` first."
```

File: tdp/cli/commands/deploy.py [hunk 11]
```diff
@@ -86,72 +38,57 @@ def deploy(
...
             for _ in deployment_iterator:
                 pass
         else:
-            session.add(deployment_iterator.log)
-            # insert pending deployment log
+            # Update deployment log to RUNNING
+            session.merge(deployment_iterator.deployment_log)
             session.commit()
-            for operation_log, service_component_log in deployment_iterator:
+            for index, (operation_log, service_component_log) in enumerate(
```
Comment (rpignolet) [inline tdp/cli/commands/deploy.py:75]: `index` is still needed ?

Comment (PaulFarault) [reply tdp/cli/commands/deploy.py]: True, I removed it.

File: tdp/cli/commands/deploy.py (hunk 10 above)
Comment (PaulFarault) [reply tdp/cli/commands/deploy.py]: Done

File: tdp/core/models/state_enum.py [hunk 12]
```diff
@@ -2,23 +2,19 @@
...
-    """
-
-    SUCCESS = "Success"
-    FAILURE = "Failure"
-    PENDING = "Pending"
 
+class _StateEnum(str, Enum):
     @classmethod
-    def has_value(cls, value):
+    def has_value(cls: Union["OperationStateEnum", "DeploymentStateEnum"], value: str):
```
Comment (rpignolet) [inline tdp/core/models/state_enum.py:17]: Why use a Union here ? You can not enumerate all `_StateEnum`'s children.
```suggestion
    def has_value(cls: _StateEnum, value: str):
```

Comment (PaulFarault) [reply tdp/core/models/state_enum.py]: Fixed

**Review by rpignolet — APPROVED** [review]

**Comment by PaulFarault** [comment]
Renamed `order` column to `operation_order` as suggested by @kpgtek

**Comment by PaulFarault** [comment]
Rebased on master

**Comment by kpgtek** [comment]
For each ```deployment_id``` in table ```operation_log```, ```operation_order``` starts with 0. If it starts with 1, perharps this could be more comprehensive.

**Comment by PaulFarault** [comment]
@kpgtek the order now start from 1!

**Comment by PaulFarault** [comment]
The browse command is not working because we are saving no-op operations in the database, which are not launched and therefore don't have any logs. I'll make a small fix for now, but it will require another PR to refactor it properly as it includes other parts of the code.

**Review by kpgtek — APPROVED** [review]

File: tdp/core/deployment/deployment_iterator.py [hunk 13]
```diff
@@ -42,21 +38,26 @@ def __init__(
...
+        self.deployment_log = log
         self._operations = operations
         self._run_operation = run_method
         self._cluster_variables = cluster_variables
-        self._iter = iter(self._operations)
+        operations_with_log = []
+        for i in range(len(self._operations)):
+            operations_with_log.append(
+                (self._operations[i], self.deployment_log.operations[i])
+            )
```
Comment (rpignolet) [inline tdp/core/deployment/deployment_iterator.py:49]:
```suggestion
        operations_with_log = [(self._operations[i], self.deployment_log.operations[i]) for i in range(len(self._operations))]
```
```

File: tdp/core/deployment/deployment_plan.py [hunk 14]
```diff
@@ -45,9 +52,29 @@ class DeploymentPlan:
...
-        self.deployment_args = deployment_args
+    def __init__(self, operations: List[Operation], deployment_log: DeploymentLog):
+        self._operations = operations
+        self._deployment_log = deployment_log
+        self._deployment_log.operations = []
+        for count, operation in enumerate(self.operations):
+            operation_log = OperationLog(
+                state=OperationStateEnum.PLANNED,
+                operation=operation.name,
+                operation_order=count + 1, # start at 1
```
[the remark this answers is not in the document]
Comment (PaulFarault) [reply tdp/core/deployment/deployment_plan.py:63]: Yep that's clearer / cleaner thx!

**Review by rpignolet — APPROVED** [review]

**Pull request merged**